Allow components to call for creation of sub components

cmacmackin opened the pull request:
This PR adds a virtual method to Component classes which lists the names and types of additional components which they might need. This information is used by the ComponentScheduler class to create these additional components automatically. Users can override default values for the configurations of these additional components by adding a named section to their config file, as usual. A `BraginskiiClosure` component has been added, which does nothing except call for the creation of all the other Braginskii components.

Closes #386

cmacmackin committed:
Implemented class to handle access permissions for options

cmacmackin committed:
Add wrapper class to control access to Options

cmacmackin committed:
Apply proper formatting

cmacmackin committed:
Switch transform method to using GuardedOptions

Note that components still aren't setting up any permissions, so there
would be runtime errors when executing the transform
methods. Furthermore, there are some missing methods I realise I need
for GuardedOptions. The unit tests are also failing to compile,
although I don't understand what the problem is for them, yet.

cmacmackin committed:
Switch transform method to using GuardedOptions

Note that components still aren't setting up any permissions, so there
would be runtime errors when executing the transform
methods. Furthermore, there are some missing methods I realise I need
for GuardedOptions. The unit tests are also failing to compile,
although I don't understand what the problem is for them, yet.

cmacmackin committed:
Added support for a permission to only read variable if already set

cmacmackin committed:
Add convenience functions for setting permissions on components

cmacmackin committed:
Fixed broken tests

cmacmackin committed:
Specify which variables are read/written for reaction components

cmacmackin committed:
Made convenience functions more flexible

cmacmackin committed:
Substitute for all_species

cmacmackin committed:
Fix access permissions in more components and update tests

cmacmackin committed:
Got everything compiling and linking. Tests still fail.

Also need to revert a bunch of changes from
Options::isSet(std::string) to IS_SET, to avoid creating objects we
don't need.

cmacmackin committed:
Reverted a bunch of uses of IS_SET()

cmacmackin committed:
Add access permissions for Braginskii closure

cmacmackin committed:
Add access permissions for remaining unit-tested components

cmacmackin committed:
Add permission information to components for time-evolution

cmacmackin committed:
Added access permissions for remaining components

All unit and integration tests pass, but many components are
untested. There may be errors in their permissions.

cmacmackin committed:
Only perform access permission checks if CHECKLEVEL >= 1

cmacmackin committed:
Added finer-grained control for which species are read/written

cmacmackin committed:
Substitute for electrons in permission data

Also added an additional constructor for SpeciesInfo

cmacmackin committed:
Updated all components to use more precise permissions

cmacmackin committed:
Try to pass GuardedOptions by value in more places

cmacmackin committed:
Switched to using class enums for permission information

cmacmackin committed:
Fixed various issues flagged in review

cmacmackin committed:
Allow permission objects to be streamed and stored in Options

cmacmackin committed:
Added unit test for topological sorting of components

cmacmackin committed:
Add ability to order components automatically

cmacmackin committed:
Fixed parsing of Permissions for variables containing colons

cmacmackin committed:
Ensure ordering of components accounts for pre-set state variables

cmacmackin committed:
Added support for sorting components based on section permissions

cmacmackin committed:
Fixed some permissions in components

cmacmackin committed:
Updated docs to describe Permissions/GuardedOptions

cmacmackin committed:
Mention new sorting and remove references to component ordering

cmacmackin committed:
Check results are the same regardless of component order

cmacmackin committed:
Check only one component can make final write

cmacmackin committed:
Fixed some incorrect access controls causing errors

cmacmackin committed:
Apply suggestions from code review

Co-authored-by: Peter Hill <peter.hill@york.ac.uk>

cmacmackin committed:
Make component variable access permissions private

cmacmackin committed:
Various minor fixes requested by @ZedThree

cmacmackin committed:
Simplified checking for nullptr

cmacmackin committed:
Rename Component::setAccess

cmacmackin committed:
Set names of variables with read-permission

cmacmackin committed:
Provide formatter to handle writing of Permissions

cmacmackin committed:
Fixed various compiler warnings

cmacmackin committed:
Applied recommendations from clang-tidy

cmacmackin committed:
Add ability for components to request additional components

cmacmackin committed:
Add Braginskii meta-component to activate all parts of closure

cmacmackin committed:
Check for missed substitutions in variable permissions

cmacmackin committed:
Fixed typo

cmacmackin committed:
Add documentation for creating sub-components.

cmacmackin committed:
Add better comments explaining component sorting process

ZedThree committed:
Use some template black magic to avoid duplication

- use forwarding reference `T&&` to avoid duplicating overloads on
  _l-values_/_r-values_
- use `std::enable_if` to hide templated functions from all `T` except
  `GuardedOptions`
- use `decltype(auto)` and `return std::forward` to ensure return type is the
  same value category as input argument

ZedThree commented on include/component.hxx:
```diff
@@ -10,43 +10,159 @@
+  ComponentInformation(const std::string& name_, const std::string& type_) : name(name_), type(type_) {}
+
+  ComponentInformation(std::string&& name_, std::string&& type_) : name(std::move(name_)), type(std::move(type_)) {}
+
```
We should be able to rely on aggregate intialisation here, and so no need for constructors:

```suggestion
```

ZedThree commented on include/component.hxx:
```diff
@@ -10,43 +10,159 @@
+         Options&
+             options, // Component settings: options[name] are specific to this component
+         Solver* solver); // Time integration solver
+
```
Formatting's going funky due to the inline comments -- I know you've not touched them, just applied formatting, but may as well remove them given we have the same info immediately above!

cmacmackin committed:
Address issues raised by @mikekryjak in review

ZedThree commented on src/component_scheduler.cxx:
```diff
@@ -32,12 +313,31 @@ ComponentScheduler::ComponentScheduler(Options &scheduler_options,
+    }
+    components.push_back(std::move(comp));
+    created_components.insert(required_components.extract(it));
+  }
```
This bit could use some commenting!

ZedThree reviewed (commented):
I've tried to review the bits that I think have been modified for this feature.

My main comment is another architectural one, and we discussed this before: using a virtual method on `Component` means that we have to create a component before discovering if it creates any others. This basically takes those additional components out of the topological sort.

I guess in some ways this is YAGNI, but perhaps this is something we do need to consider?

cmacmackin committed:
Merge pull request #447 from boutproject/access-control-template-magic

Use some template black magic to avoid duplication

cmacmackin commented:
> I've tried to review the bits that I think have been modified for this feature.
> 
> My main comment is another architectural one, and we discussed this before: using a virtual method on `Component` means that we have to create a component before discovering if it creates any others. This basically takes those additional components out of the topological sort.
> 
> I guess in some ways this is YAGNI, but perhaps this is something we do need to consider?

This shouldn't take the new components out of the topological sort. The sorting only happens after all components have been created, so adding more components won't affect that.

cmacmackin commented on src/component_scheduler.cxx:
```diff
@@ -1,7 +1,265 @@
+/// topolgically sorted to ensure variables are written and read in
+/// the right order.
+std::vector<std::unique_ptr<Component>>
+sortComponents(std::vector<std::unique_ptr<Component>>&& components) {
```
I could do that, although the sorting doesn't actually happen in-place per-se. I construct a new std::vector by moving elements from the first. Basically, the question boils down to whether we overwrite the old (now invalid) `ComponentScheduler::components` vector within `sortComponents` or after we call it (as currently happens on [line 340](https://github.com/boutproject/hermes-3/pull/445/files#diff-3539d3bec156fbfbca693b2bf24246d4f38ed24a307d4e8327bea217d8e7d697R340)). I have a weird bias towards using return values but in this case I can see that it probably makes more sense to modify the reference argument within the function.

cmacmackin committed:
Responded to some comments by @ZedThree

The comments were in PR #445, but some of them actually relate to work
done in #428. Only the latter are dealt with in this PR.

bendudson commented on src/component_scheduler.cxx:
```diff
@@ -20,6 +281,26 @@ ComponentScheduler::ComponentScheduler(Options &scheduler_options,
+      electrons.push_back(name_trimmed);
+    }
+    // FIXME: Would there be any spcies without AA? Is there any other
+    // reliable way to identify what is a species?
```
I think there are good reasons for not making species a first-class entity, as models may want to evolve things like groups of species (bundled charge states), different vibrational states, or energy groups (e.g. beam electrons), as distinct 'species', so what is meant by a species is intentionally flexible.

One possibility is to have the species 'self identify': Each species currently sets a 'charge' and 'AA' (mass) property in the state. We could just add a 'type' property that determines how it should be treated in e.g. collisions.

cmacmackin replied:
I suppose another option would be to have an additional virtual method on components:
```
virtual std::optional<std::string> species() const { return {}; }
```
If the component evolves a species then it would override the method to return that species' name (which is usually the same as the component name).

cmacmackin committed:
Add missed variable permission

cmacmackin commented on include/component.hxx:
```diff
@@ -10,43 +10,159 @@
+                     const std::vector<std::string>& neutrals,
+                     const std::vector<std::string>& positive_ions,
+                     const std::vector<std::string> & negative_ions)
+    : electrons(electrons), neutrals(neutrals), positive_ions(positive_ions), negative_ions(negative_ions), ions(positive_ions) {
```
Addressed in #428.

cmacmackin commented on src/component_scheduler.cxx:
```diff
@@ -1,7 +1,265 @@
+                                      PermissionTypes::Write),
+                                  final_writes = getPermissionComponentMap(
+                                      components, variable_hierarchy,
+                                      PermissionTypes::Final);
```
Addressed in #428.

cmacmackin committed:
Reordered/changed heading of docs so reads better as introduction

cmacmackin committed:
Allow integration tests to be run in parallel in CI

cmacmackin committed:
Ensure reactions do not depend on order in which they are built

cmacmackin committed:
Fix issues when compiling with CHECKLEVEL 0

- Unused arguments in some of the functions for GuardedOptions
- Unit tests expecting an exception to be thrown, but this only
  happens when checking is done.

cmacmackin committed:
Revert "Allow integration tests to be run in parallel in CI"

This reverts commit c5bd5898e48b497337fba06b0474e897eeb2c7be, as
running integration tests in parallel on the CI machine proved to be
slower. Perhaps it is memory-bounded?

cmacmackin commented:
The `1D-recycling-dthe` integration test is failing in CI, with relative differences on the order of >1e-2. (Far too high!) This appears to be due to the radically different order in which components end up being built, which is a side-effect of some of the changes in the PR. Previously, the sorting algorithm used the order given in the input file as its starting point. Now it basically sorts the components alphabetically by name and type before it does the topological sort. This means that the final order of the components after topological sorting is now radically different from what is given in the input file.

I've tried rewriting it slightly so that it will go back to using the order in the input file as its starting point for topological sorting. This makes the test pass, but I'm still concerned by what I'm seeing. It would appear that the permission information I am currently specifying on some components is not adequate to guarantee the same result. Perhaps something hasn't been marked as the final-write that should be. I note that this is the only integration test which has had this problem; all the others worked fine even with a radical reordering of the components. That might give us something to go on.

cmacmackin commented:
Here is the order the components are executed when I get the error described above:


- d (evolve_density)
- d (evolve_pressure)
- d+ (evolve_density)
- d+ (evolve_pressure)
- he (evolve_density)
- he+ (evolve_density)
- t (evolve_density)
- t+ (evolve_density)
- e (quasineutral)
- e (evolve_pressure)
- he (evolve_pressure)
- he+ (evolve_pressure)
- t (evolve_pressure)
- t+ (evolve_pressure)
- braginskii_collisions (braginskii_collisions)
- d (evolve_momentum)
- d (noflow_boundary)
- he (evolve_momentum)
- he (noflow_boundary)
- d+ (evolve_momentum)
- d+ (noflow_boundary)
- he+ (evolve_momentum)
- t (evolve_momentum)
- t+ (evolve_momentum)
- e (zero_current)
- e (noflow_boundary)
- he+ (noflow_boundary)
- t+ (noflow_boundary)
- sheath_boundary (sheath_boundary)
- reactions (d + d+ -> d+ + d)
- t (noflow_boundary)
- reactions (d + t+ -> d+ + t)
- reactions (t + d+ -> t+ + d)
- reactions (t + t+ -> t+ + t)
- braginskii_conduction (braginskii_conduction)
- braginskii_ion_viscosity (braginskii_ion_viscosity)
- braginskii_thermal_force (braginskii_thermal_force)
- d+ (upstream_density_feedback)
- reactions (d + e -> d+ + 2e)
- reactions (d+ + e -> d)
- reactions (he + e -> he+ + 2e)
- reactions (he+ + e -> he)
- reactions (t + e -> t+ + 2e)
- reactions (t+ + e -> t)
- electron_force_balance (electron_force_balance)
- he+ (upstream_density_feedback)
- neutral_parallel_diffusion (neutral_parallel_diffusion)
- recycling (recycling)
- t+ (upstream_density_feedback)

@mikekryjak, @bendudson is there anything in this ordering that stands out to you as wrong?

cmacmackin committed:
Add ability to print component order and deactivate sorting

cmacmackin committed:
Fixed some minor permission issues

cmacmackin commented:
I've worked out the two components that are causing problems (although not why): `braginskii_ion_viscosity` and `upstream_density_feedback`. It seems the latter has to be executed first, even though it doesn't appear to touch anything that would conflict with the viscosity.
